File: backend/app/schemas/subject.py
import uuid
from datetime import datetime

from pydantic import ConfigDict, Field, model_validator

from app.schemas.common import NonEmptyUpdateModel, StrippedModel


class SubjectBase(StrippedModel):
    name: str = Field(min_length=2, max_length=150)
    code: str = Field(min_length=2, max_length=20)
    credits: int = Field(ge=1, le=10, default=3)
    theory_hours_per_week: int = Field(ge=0, le=20, default=0)
    practical_hours_per_week: int = Field(ge=0, le=20, default=0)
    semester_id: uuid.UUID
    department_id: uuid.UUID

    @model_validator(mode="after")
    def validate_has_teaching_hours(self) -> "SubjectBase":
        if self.theory_hours_per_week + self.practical_hours_per_week <= 0:
            raise ValueError("A subject must have at least one theory or practical hour per week")
        return self


class SubjectCreate(SubjectBase):
    pass


class SubjectUpdate(NonEmptyUpdateModel):
    name: str | None = Field(default=None, min_length=2, max_length=150)
    code: str | None = Field(default=None, min_length=2, max_length=20)
    credits: int | None = Field(default=None, ge=1, le=10)
    theory_hours_per_week: int | None = Field(default=None, ge=0, le=20)
    practical_hours_per_week: int | None = Field(default=None, ge=0, le=20)
    semester_id: uuid.UUID | None = None
    department_id: uuid.UUID | None = None


class SubjectRead(SubjectBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

File: frontend/lib/validators/subject.ts
import { z } from "zod";

import { trimmedString } from "@/lib/validators/shared";

export const subjectFormSchema = z
  .object({
    name: trimmedString().min(2, "Subject name is required").max(150),
    code: trimmedString().min(2, "Subject code is required").max(20),
    credits: z.coerce.number().int().min(1).max(10),
    theory_hours_per_week: z.coerce.number().int().min(0).max(20),
    practical_hours_per_week: z.coerce.number().int().min(0).max(20),
    semester_id: z.string().uuid("Select a semester"),
    department_id: z.string().uuid("Select a department"),
  })
  .refine((value) => value.theory_hours_per_week + value.practical_hours_per_week > 0, {
    message: "Add at least one theory or practical hour per week",
    path: ["theory_hours_per_week"],
  });

export type SubjectFormValues = z.infer<typeof subjectFormSchema>;
